### PR TITLE
Add file handlers with tests

### PR DESCRIPTION
**Pull request by kirillsakh**
Add file handlers (.json and .npy) with tests.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -3,15 +3,15 @@ repos:
     rev: v1.10.0
     hooks:
       - id: mypy
-        args: [--strict, --ignore-missing-imports]
+        args: [--ignore-missing-imports, --explicit-package-bases]
```
Comment (tynaneu) [inline .pre-commit-config.yaml:6]: this change is not related to the commit - it should be in separate commit (with description why it is done)

Comment (tynaneu) [reply .pre-commit-config.yaml:6]: please do not use "explicit-package-bases" - you should use `__init__.py` files to initialize packages

Comment (tynaneu) [reply .pre-commit-config.yaml:6]: why you removed "strict" option

File: src/file_handlers/base_handler.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import numpy as np
+
+
+class FileFormatHandler:
```
Comment (tynaneu) [inline src/file_handlers/base_handler.py:4]: interfaces in Python should inherit from ABC or Protocol

File: tests/file_handlers/conftest.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+
+@pytest.fixture
+def numpy_handler() -> NumpyHandler:
+    return NumpyHandler()
+
+
+def ndarray(num_samples: int, num_of_dimensions: int) -> np.ndarray:
+    """Fixture to generate N-dimensional arrays."""
+    feature_vector_sizes = np.random.randint(1, 6, size=num_of_dimensions - 1)
```
Comment (tynaneu) [inline tests/file_handlers/conftest.py:31]: maybe there was some misunderstanding or mistake from our side, but the input array should have shape: (num_samples, vector_space_dimension) - clustering is done on n-dimensional vectors,
you can simply use: `np.random.normal(0,1,(num_samples, dimension))` (there is no need for this helper function)

File: tests/file_handlers/conftest.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    shape = (num_samples,) + tuple(feature_vector_sizes)
+    return np.random.randn(*shape)
+
+
+@pytest.fixture
+def array_4_d() -> np.ndarray:
+    return ndarray(num_samples=300, num_of_dimensions=4)
```
Comment (tynaneu) [inline tests/file_handlers/conftest.py:38]: use np.random.normal right here (see above comment)

File: tests/file_handlers/test_base_handler.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import numpy as np
+import pytest
+
+from src.file_handlers.base_handler import FileFormatHandler
+
+
+class TestFileFormatHandler:
```
Comment (tynaneu) [inline tests/file_handlers/test_base_handler.py:7]: there is no need to test this abstract class behaviour (use python ABC or Protocol)

File: tests/file_handlers/test_json_handler.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import numpy as np
+
+from src.file_handlers.json_handler import JsonHandler
+
+
+def test_load(
+    json_handler: JsonHandler, array_4_d: np.ndarray, json_file_path: str
+) -> None:
+    loaded_data = json_handler.load(json_file_path)
+    assert np.array_equal(loaded_data, array_4_d)
```
Comment (tynaneu) [inline tests/file_handlers/test_json_handler.py:13]: please use `np.testing.assert_array_equal`

File: tests/file_handlers/test_json_handler.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+
+def test_save(json_handler: JsonHandler, tmpdir: Path) -> None:
+    data = np.array([1, 2, 3])
```
Comment (tynaneu) [inline tests/file_handlers/test_json_handler.py:17]: why do you create new data if you already have a fixture for it with the correct shape?

File: tests/file_handlers/test_numpy_handler.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import numpy as np
+
+from src.file_handlers.numpy_handler import NumpyHandler
+
+
+def test_load(
+    numpy_handler: NumpyHandler, array_4_d: np.ndarray, numpy_file_path: str
+) -> None:
+    loaded_data = numpy_handler.load(numpy_file_path)
+    assert np.array_equal(loaded_data, array_4_d)
```
Comment (tynaneu) [inline tests/file_handlers/test_numpy_handler.py:12]: please use np.testing

File: tests/file_handlers/test_numpy_handler.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+
+def test_save(numpy_handler: NumpyHandler, tmpdir: Path) -> None:
+    data = np.array([1, 2, 3])
```
Comment (tynaneu) [inline tests/file_handlers/test_numpy_handler.py:16]: please do not create data, use array_4_d

File: src/file_handlers/base_handler.py (hunk 2 above)
Comment (kirillsakh) [reply src/file_handlers/base_handler.py:4]: I will address this issue. Previously, I was aware that within Python, there exist two primary options for defining interfaces:

- Abstract Base Classes (ABC)

- Basic classes employing NotImplementedError.

After conducting a brief investigation, it became evident that abstract classes offer advantages over the latter method.

File: tests/file_handlers/conftest.py (hunk 3 above)
Comment (kirillsakh) [reply tests/file_handlers/conftest.py:31]: I agree, evidently there was a misunderstanding on my part. I interpreted `Vstupni matice bude mit patrne vice dimenzi naz jen dve. Ale ano prvni dimenze budou vzorky, dalsi dimenze bude vektor. tedy N x M1 x M2 ... x Mx` incorrectly.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (kirillsakh) [reply .pre-commit-config.yaml:6]: Appreciate the heads-up. I'll need to look into the reason behind this change.

File: tests/file_handlers/test_json_handler.py (hunk 6 above)
Comment (kirillsakh) [reply tests/file_handlers/test_json_handler.py:13]: I wasn't aware that `np.testing.assert_array_equal` existed. Thanks for the suggestion!

File: .pre-commit-config.yaml (hunk 1 above)
Comment (kirillsakh) [reply .pre-commit-config.yaml:6]: Here's the rationale behind the modifications:

Screenshot 2024-05-10 at 11 13 09

Comment (kirillsakh) [reply .pre-commit-config.yaml:6]: Screenshot 2024-05-10 at 11 24 14

Comment (kirillsakh) [reply .pre-commit-config.yaml:6]: after modification:
Screenshot 2024-05-10 at 11 25 20

File: tests/file_handlers/test_json_handler.py (hunk 7 above)
Comment (kirillsakh) [reply tests/file_handlers/test_json_handler.py:17]: My intention was to design this method to be universal across any specific implementation. It should be capable of handling any ndarray objects, hence the decision not to rely on a single fixture for all methods.
